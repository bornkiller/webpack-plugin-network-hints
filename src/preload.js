/**
 * @description - calculate preload link attributes
 * @author - huang.jian <hjj491229492@hotmail.com>
 */

// packages
import { extname } from 'path';
import { lookup } from 'mime-types';

// scope
const scripts = ['.js', '.mjs'];
const stylesheets = ['.css'];
const fonts = ['.woff', '.woff2', '.ttf', '.eot'];
const images = ['.jpg', '.jpeg', '.png', '.gif', '.webp'];
const documents = ['.html'];

/**
 * @description - compose specific attrs for specific files
 *
 * @param {string} asset
 */
function preloadSpecificAttrs(asset) {
  const extension = extname(asset);

  switch (true) {
    case scripts.includes(extension):
      return { as: 'script' };
    case stylesheets.includes(extension):
      return { as: 'stylesheet', onload: `this.rel='stylesheet'` };
    case fonts.includes(extension):
      return { as: 'font', crossorigin: 'anonymous' };
    case images.includes(extension):
      return { as: 'image' };
    case documents.includes(extension):
      return { as: 'document' };
    default:
      return {};
  }
}

/**
 * @param {string} asset
 *
 * @return {object}
 */
export default function preload(asset) {
  const attrs = {
    rel: 'preload',
    href: asset,
    type: lookup(asset),
  };
  const specifics = preloadSpecificAttrs(asset);

  return { ...attrs, ...specifics };
}
