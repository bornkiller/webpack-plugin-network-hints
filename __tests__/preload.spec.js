/**
 * @description - tools unit tests
 * @author - huang.jian <hjj491229492@hotmail.com>
 */

// packages
import preload from '../src/preload';

describe('tools suits', () => {
  it('should preload scripts', () => {
    expect(preload('jest.chunk.js')).toMatchSnapshot();
  });

  it('should preload stylesheets', () => {
    expect(preload('jest.chunk.css')).toMatchSnapshot();
  });

  it('should preload fonts', () => {
    expect(preload('jest.chunk.woff2')).toMatchSnapshot();
  });

  it('should preload images', () => {
    expect(preload('jest.chunk.jpg')).toMatchSnapshot();
  });
  it('should preload document', () => {
    expect(preload('jest.chunk.html')).toMatchSnapshot();
  });
});
