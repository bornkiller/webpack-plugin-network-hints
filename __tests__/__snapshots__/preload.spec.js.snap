// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tools suits should preload document 1`] = `
Object {
  "as": "document",
  "href": "jest.chunk.html",
  "rel": "preload",
  "type": "text/html",
}
`;

exports[`tools suits should preload fonts 1`] = `
Object {
  "as": "font",
  "crossorigin": "anonymous",
  "href": "jest.chunk.woff2",
  "rel": "preload",
  "type": "font/woff2",
}
`;

exports[`tools suits should preload images 1`] = `
Object {
  "as": "image",
  "href": "jest.chunk.jpg",
  "rel": "preload",
  "type": "image/jpeg",
}
`;

exports[`tools suits should preload scripts 1`] = `
Object {
  "as": "script",
  "href": "jest.chunk.js",
  "rel": "preload",
  "type": "application/javascript",
}
`;

exports[`tools suits should preload stylesheets 1`] = `
Object {
  "as": "stylesheet",
  "href": "jest.chunk.css",
  "onload": "this.rel='stylesheet'",
  "rel": "preload",
  "type": "text/css",
}
`;
